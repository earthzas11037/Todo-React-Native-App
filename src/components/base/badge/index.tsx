import { badgeTv } from '@/components/base/badge/badge.style'
import { Text } from '@/components/base/text'
import React from 'react'
import { View } from 'react-native'
import { tv, type VariantProps } from 'tailwind-variants'

interface BadgeProps extends VariantProps<typeof badgeTv> {
  content?: string | number
  invisible?: boolean
  children?: React.ReactNode
  className?: string
  badgeClassName?: string
  contentClassName?: string
}

export const Badge: React.FC<BadgeProps> = ({
  children,
  content,
  size = 'medium',
  color = 'primary',
  variant = 'standard',
  invisible = false,
  className,
  badgeClassName,
  contentClassName
}) => {
  const styles = React.useMemo(() => badgeTv({ size, color, variant, invisible }), [size, color, variant, invisible])
  const displayContent =
    typeof content === 'number' && variant !== 'dot' ? (content > 99 ? '99+' : content.toString()) : content

  return (
    <View className={`${styles.container({ className })}`}>
      {children}
      {!!content && (
        <View className={`${styles.badge({ className: badgeClassName })}`}>
          <Text className={`${styles.content({ className: contentClassName })}`}>
            {variant !== 'dot' ? displayContent : ''}
          </Text>
        </View>
      )}
    </View>
  )
}
