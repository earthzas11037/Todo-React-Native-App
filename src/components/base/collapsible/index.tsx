import React from 'react'
import { cssInterop } from 'nativewind'
import LibCollapsible, { CollapsibleProps as LibCollapsibleProps } from 'react-native-collapsible'

// Apply NativeWind integration
cssInterop(LibCollapsible, { className: 'style' })

interface CollapsibleProps extends LibCollapsibleProps {
  /**
   * Additional styles for the container
   */
  className?: string
  expanded: boolean
}

/**
 * Reusable Collapsible component with default props
 */
export const Collapsible: React.FC<CollapsibleProps> = ({ expanded, duration = 500, children, ...props }) => {
  return (
    <LibCollapsible collapsed={!expanded} duration={duration} {...props}>
      {children}
    </LibCollapsible>
  )
}
