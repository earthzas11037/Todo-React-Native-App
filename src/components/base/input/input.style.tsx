import { tv } from 'tailwind-variants'

export const inputTv = tv({
  slots: {
    container: '',
    label: 'text-text-primary mb-1 text-base',
    inputContainer: 'rounded-lg bg-white px-4 py-2 min-h-[50px] flex flex-row',
    input: 'flex-1 font-notoSansThai text-[14px] font-normal leading-normal',
    helperText: 'text-base text-error-500 mt-1'
  },
  variants: {
    focused: {
      true: {
        inputContainer: 'border-primary-500 border'
      },
      false: {
        inputContainer: 'shadow-none border border-neutral-300'
      }
    },
    color: {
      primary: {
        label: '',
        inputContainer: '',
        helperText: ''
      },
      secondary: {
        label: 'text-secondary-500',
        inputContainer: 'border-secondary-500',
        helperText: 'text-secondary-500'
      },
      error: {
        label: 'text-error-500',
        inputContainer: 'border-error-500',
        helperText: 'text-error-500'
      },
      info: {
        label: 'text-info-500',
        inputContainer: 'border-info-500',
        helperText: 'text-info-500'
      },
      success: {
        label: 'text-success-500',
        inputContainer: 'border-success-500',
        helperText: 'text-success-500'
      },
      warning: {
        label: 'text-warning-500',
        inputContainer: 'border-warning-500',
        helperText: 'text-warning-500'
      }
    },
    error: {
      true: {
        inputContainer: 'border-error-500',
        label: 'text-error-500'
      }
    },
    disabled: {
      true: {
        label: 'text-text-disabled',
        inputContainer: 'bg-action-hover text-text-disabled border-outline-border',
        input: 'text-text-disabled',
        helperText: 'text-text-disabled'
      }
    }
  },
  compoundVariants: [
    {
      color: ['secondary', 'error', 'info', 'success', 'warning'],
      focused: true,
      className: {
        inputContainer: 'shadow-none'
      }
    }
  ],
  defaultVariants: {
    focused: false,
    error: false,
    disabled: false
  }
})
