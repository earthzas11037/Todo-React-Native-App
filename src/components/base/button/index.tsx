import React, { JSX } from 'react'
import type { TouchableWithoutFeedbackProps } from 'react-native'
import { ActivityIndicator, TouchableWithoutFeedback, View } from 'react-native'
import type { VariantProps } from 'tailwind-variants'
import colors from '@/config/theme/colors'
import { buttonTv } from '@/components/base/button/button.style'
import { twMerge } from 'tailwind-merge'
import { LinearGradient } from '@/components/base/linear-gradient'
import { Text } from '@/components/base/text'

type ButtonVariants = VariantProps<typeof buttonTv>
interface Props extends ButtonVariants, Omit<TouchableWithoutFeedbackProps, 'disabled'> {
  label?: string
  loading?: boolean
  startIcon?: JSX.Element
  endIcon?: JSX.Element
  className?: string
  textClassName?: string
  iconClassName?: string
}

export const Button = React.forwardRef<View, Props>(
  (
    {
      label,
      loading = false,
      variant = 'contained',
      size = 'medium',
      color = 'primary',
      disabled = false,
      fullWidth = false,
      className = '',
      testID,
      textClassName = '',
      iconClassName = '',
      startIcon,
      endIcon,
      ...props
    },
    ref
  ) => {
    const [pressed, setPressed] = React.useState(false)

    const styles = React.useMemo(
      () => buttonTv({ variant, size, color, disabled, fullWidth }),
      [variant, size, color, disabled, fullWidth]
    )

    const isGradient = colors.gradient.primary && color === 'primary' && variant === 'contained' && disabled === false
    const gradientColor = colors.gradient.primary

    const handlePressIn = () => setPressed(true)
    const handlePressOut = () => setPressed(false)

    const dynamicStyle = pressed ? 'opacity-75' : ''

    if (isGradient) {
      return (
        <TouchableWithoutFeedback
          disabled={disabled || loading}
          {...props}
          testID={testID}
          onPressIn={handlePressIn}
          onPressOut={handlePressOut}
        >
          <LinearGradient
            ref={ref}
            {...(gradientColor as any)}
            className={twMerge(`${styles.container({ className })}`, dynamicStyle)}
          >
            {loading ? (
              <ActivityIndicator
                size='small'
                color='white'
                testID={testID ? `${testID}-activity-indicator` : undefined}
              />
            ) : (
              <>
                {startIcon &&
                  React.cloneElement(startIcon as any, {
                    className: `mr-[10px] ${styles.icon({ className: iconClassName })} ${
                      startIcon.props.className || ''
                    }`
                  })}
                <Text
                  testID={testID ? `${testID}-label` : undefined}
                  className={styles.label({ className: textClassName })}
                >
                  {label}
                </Text>
                {endIcon &&
                  React.cloneElement(endIcon as any, {
                    className: `ml-[10px] ${styles.icon({ className: iconClassName })} ${endIcon.props.className || ''}`
                  })}
              </>
            )}
          </LinearGradient>
        </TouchableWithoutFeedback>
      )
    }
    return (
      <TouchableWithoutFeedback
        disabled={disabled || loading}
        {...props}
        // ref={ref}
        testID={testID}
        onPressIn={handlePressIn}
        onPressOut={handlePressOut}
      >
        <View ref={ref} className={`${styles.container({ className })} ${dynamicStyle}`}>
          {loading ? (
            <ActivityIndicator
              size='small'
              className={styles.indicator()}
              testID={testID ? `${testID}-activity-indicator` : undefined}
            />
          ) : (
            <>
              {startIcon &&
                React.cloneElement(startIcon as any, {
                  className: `mr-[10px] ${styles.icon({ className: iconClassName })} ${startIcon.props.className || ''}`
                })}
              <Text
                testID={testID ? `${testID}-label` : undefined}
                className={styles.label({ className: textClassName })}
              >
                {label}
              </Text>
              {endIcon &&
                React.cloneElement(endIcon as any, {
                  className: `ml-[10px] ${styles.icon({ className: iconClassName })} ${endIcon.props.className || ''}`
                })}
            </>
          )}
        </View>
      </TouchableWithoutFeedback>
    )
  }
)
