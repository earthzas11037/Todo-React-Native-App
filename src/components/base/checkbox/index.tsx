import { MotiView } from 'moti'
import React, { useCallback, useEffect, useRef } from 'react'
import { Animated, I18nManager, Pressable, type PressableProps, View } from 'react-native'
import colors from '@/config/theme/colors'
import { Text } from '@/components/base/text'
import { VariantProps } from 'tailwind-variants'
import CustomIcon from '@/components/base/icon'
import { checkboxTv } from '@/components/base/checkbox/checkbox.style'

const SIZE = 18
const WIDTH = 50
const HEIGHT = 28
const THUMB_HEIGHT = 22
const THUMB_WIDTH = 22
const THUMB_OFFSET = 4

type CheckboxVariants = VariantProps<typeof checkboxTv>

export interface RootProps extends CheckboxVariants, Omit<PressableProps, 'onPress'> {
  onChange: (checked: boolean) => void
  checked?: boolean
  className?: string
  accessibilityLabel?: string
}

export type IconProps = {
  checked: boolean
  disabled?: boolean | null | undefined
  color?: 'primary' | 'secondary' | 'error' | 'info' | 'success' | 'warning'
}

export const Root = ({ checked = false, children, onChange, disabled, className = '', ...props }: RootProps) => {
  const handleChange = useCallback(() => {
    onChange(!checked)
  }, [onChange, checked])

  return (
    <Pressable
      onPress={handleChange}
      className={`flex-row items-center ${className} ${disabled ? 'opacity-50' : ''}`}
      accessibilityState={{ checked }}
      disabled={disabled}
      {...props}
    >
      {children}
    </Pressable>
  )
}

type LabelProps = {
  text: string
  className?: string
  testID?: string
}

const Label = ({ text, testID, className = '' }: LabelProps) => {
  return (
    <Text testID={testID} className={` ${className} pl-[6px]`}>
      {text}
    </Text>
  )
}

export const CheckboxIcon = ({ checked = false, color = 'primary', disabled }: IconProps) => {
  const _color = checked ? colors[color][500] : colors.text.disabled
  return (
    <MotiView
      style={{
        height: SIZE,
        width: SIZE,
        borderColor: _color,
        shadowColor: colors.primary[500],
        shadowOffset: { width: 0, height: 2 },
        shadowOpacity: checked ? 0.25 : 0,
        shadowRadius: 3.84,
        elevation: 5
      }}
      className='items-center justify-center rounded-[5px] border-2'
      from={{ backgroundColor: 'transparent', borderColor: '#CCCFD6' }}
      animate={{
        backgroundColor: checked ? _color : 'transparent',
        borderColor: _color
      }}
      transition={{
        backgroundColor: { type: 'timing', duration: 100 },
        borderColor: { type: 'timing', duration: 100 }
      }}
    >
      <MotiView
        from={{ opacity: 0 }}
        animate={{ opacity: checked ? 1 : 0 }}
        transition={{ opacity: { type: 'timing', duration: 100 } }}
      >
        <CustomIcon name='IconCheck' size={16} color='white' />
      </MotiView>
    </MotiView>
  )
}

const CheckboxRoot = ({ checked = false, children, ...props }: RootProps) => {
  return (
    <Root checked={checked} accessibilityRole='checkbox' {...props}>
      {children}
    </Root>
  )
}

const CheckboxBase = ({
  checked = false,
  testID,
  label,
  color = 'primary',
  ...props
}: RootProps & { label?: string }) => {
  return (
    <CheckboxRoot checked={checked} testID={testID} color={color} {...props}>
      <CheckboxIcon checked={checked} color={color} />
      {label ? <Label text={label} testID={testID ? `${testID}-label` : undefined} className='pr-2' /> : null}
    </CheckboxRoot>
  )
}

export const Checkbox = Object.assign(CheckboxBase, {
  Icon: CheckboxIcon,
  Root: CheckboxRoot,
  Label
})

export const RadioIcon = ({ checked = false, color = 'primary', disabled }: IconProps) => {
  const borderColor = checked ? colors[color][500] : disabled ? colors.gray.o.light : colors.text.disabled
  const innerColor = !disabled ? 'transparent' : checked ? colors[color][500] : colors.gray.o.light
  return (
    <View
      style={[
        {
          height: SIZE,
          width: SIZE,
          borderWidth: checked || (!checked && disabled) ? 5 : 2,
          justifyContent: 'center',
          alignItems: 'center',
          borderRadius: SIZE / 2,
          backgroundColor: 'transparent',
          shadowOffset: { width: 0, height: 2 },
          shadowRadius: 3.84,
          elevation: 5,
          borderColor: borderColor,
          shadowColor: disabled ? 'transparent' : colors[color][500],
          shadowOpacity: checked ? 0.25 : 0
        }
      ]}
    >
      {checked || disabled ? (
        <View
          style={[
            {
              height: 8,
              width: 8,
              borderRadius: 4
            },
            { backgroundColor: innerColor }
          ]}
        />
      ) : null}
    </View>
  )
}

const RadioRoot = ({ checked = false, children, ...props }: RootProps) => {
  return (
    <Root checked={checked} accessibilityRole='radio' {...props}>
      {children}
    </Root>
  )
}

const RadioBase = ({
  checked = false,
  testID,
  label,
  color = 'primary',
  disabled,
  ...props
}: RootProps & { label?: string }) => {
  return (
    <RadioRoot checked={checked} testID={testID} {...props}>
      <RadioIcon checked={checked} color={color} disabled={disabled} />
      {label ? <Label text={label} testID={testID ? `${testID}-label` : undefined} /> : null}
    </RadioRoot>
  )
}

export const Radio = Object.assign(RadioBase, {
  Icon: RadioIcon,
  Root: RadioRoot,
  Label
})

export const SwitchIcon = ({ checked = false }: IconProps) => {
  const translateX = useRef(new Animated.Value(checked ? THUMB_OFFSET : WIDTH - THUMB_WIDTH - THUMB_OFFSET)).current
  const backgroundColor = checked ? colors.primary[500] : colors.text.disabled

  useEffect(() => {
    Animated.timing(translateX, {
      toValue: checked ? THUMB_OFFSET : WIDTH - THUMB_WIDTH - THUMB_OFFSET,
      duration: 150,
      useNativeDriver: true
    }).start()
  }, [checked])

  return (
    <View className='w-[50px] justify-center'>
      <View className='overflow-hidden rounded-full'>
        <View
          style={{
            width: WIDTH,
            height: HEIGHT,
            backgroundColor
          }}
        />
      </View>
      <Animated.View
        style={{
          height: THUMB_HEIGHT,
          width: THUMB_WIDTH,
          position: 'absolute',
          backgroundColor: 'white',
          borderRadius: 13,
          right: 0,
          transform: [
            {
              translateX: I18nManager.isRTL ? translateX : Animated.multiply(translateX, -1)
            }
          ]
        }}
      />
    </View>
  )
}

const SwitchRoot = ({ checked = false, children, ...props }: RootProps) => {
  return (
    <Root checked={checked} accessibilityRole='switch' {...props}>
      {children}
    </Root>
  )
}

const SwitchBase = ({ checked = false, testID, label, ...props }: RootProps & { label?: string }) => {
  return (
    <SwitchRoot checked={checked} testID={testID} {...props}>
      <SwitchIcon checked={checked} />
      {label ? <Label text={label} testID={testID ? `${testID}-label` : undefined} /> : null}
    </SwitchRoot>
  )
}

export const Switch = Object.assign(SwitchBase, {
  Icon: SwitchIcon,
  Root: SwitchRoot,
  Label
})
