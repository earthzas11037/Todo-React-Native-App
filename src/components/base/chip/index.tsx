import React from 'react'
import { TouchableOpacity, Pressable } from 'react-native'
import { type VariantProps } from 'tailwind-variants'
import CustomIcon from '@/components/base/icon'
import { chipTv } from '@/components/base/chip/chip.style'
import { Text } from '@/components/base/text'

interface ChipProps extends VariantProps<typeof chipTv> {
  label: string
  icon?: JSX.Element
  deleteIcon?: JSX.Element
  onDelete?: () => void
  onPress?: () => void
  disabled?: boolean
  className?: string
  labelClassName?: string
  iconClassName?: string
  deleteIconClassName?: string
}

export const Chip: React.FC<ChipProps> = ({
  label,
  icon,
  deleteIcon,
  color = 'default',
  size = 'medium',
  disabled,
  variant = 'filled',
  fullWidth,
  onPress,
  onDelete,
  className,
  labelClassName,
  iconClassName,
  deleteIconClassName
}) => {
  const styles = React.useMemo(
    () => chipTv({ color, size, disabled, variant, fullWidth }),
    [color, size, disabled, variant, fullWidth]
  )

  return (
    <Pressable className={`${styles.container({ className })}`} onPress={onPress} disabled={disabled}>
      {icon &&
        React.cloneElement(icon as any, {
          className: `${styles.icon({ className: iconClassName })} ${icon.props.className || ''}`
        })}
      <Text className={`${styles.label({ className: labelClassName })}`}>{label}</Text>
      {onDelete && (
        <TouchableOpacity onPress={onDelete}>
          {deleteIcon ? (
            React.cloneElement(deleteIcon as any, {
              className: `${styles.deleteIcon({ className: deleteIconClassName })} ${deleteIcon.props.className || ''}`
            })
          ) : (
            <CustomIcon
              name='IconCircleX'
              size={size == 'small' ? 16 : size == 'medium' ? 20 : 24}
              className={`${styles.deleteIcon({ className: deleteIconClassName })}`}
            />
          )}
        </TouchableOpacity>
      )}
    </Pressable>
  )
}
