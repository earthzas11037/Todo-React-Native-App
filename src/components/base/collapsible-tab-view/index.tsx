import React from 'react'
import {
  Tabs,
  MaterialTabBar,
  CollapsibleProps,
  MaterialTabBarProps,
  MaterialTabItemProps,
  MaterialTabItem,
  TabItemProps
} from 'react-native-collapsible-tab-view'
import { Dimensions } from 'react-native'
import colors from '@/config/theme/colors'
import { Text } from '@/components/base/text'
import { twMerge } from 'tailwind-merge'

// Custom TabBar Component
const CustomTabBar: React.FC<MaterialTabBarProps<any>> = props => {
  const { width } = Dimensions.get('window')
  return (
    <MaterialTabBar
      {...props}
      activeColor={colors.primary[500]}
      inactiveColor={colors.text.primary}
      style={{
        backgroundColor: 'transparent'
      }}
      contentContainerStyle={{}}
      labelStyle={{}}
      tabStyle={{}}
      TabItemComponent={({ focusable, ...rest }: MaterialTabItemProps<any>) => {
        return (
          <MaterialTabItem
            {...rest}
            activeColor={colors.primary[500]}
            label={(itemProps: TabItemProps<any>) => {
              const isFocused = props.focusedTab.value == itemProps.name
              return (
                <Text
                  className={twMerge(`text-center px-5 py-2`)}
                  style={{ width: width / props.tabNames.length, color: props.activeColor }}
                >
                  {itemProps.name as String}
                </Text>
              )
            }}
          />
        )
      }}
      indicatorStyle={{
        backgroundColor: colors.primary[500],
        shadowOpacity: 0
      }}
      scrollEnabled
    />
  )
}

// Custom Tabs.Container with default TabBar customization
const CustomTabsContainer: React.FC<CollapsibleProps> = props => {
  return (
    <Tabs.Container
      {...props}
      renderTabBar={tabBarProps => <CustomTabBar {...tabBarProps} />}
      headerContainerStyle={{ shadowOpacity: 0, borderColor: colors.divider, borderBottomWidth: 1 }}
    />
  )
}

// Re-export Tabs components
const CustomTabs = {
  Container: CustomTabsContainer,
  Tab: Tabs.Tab,
  ScrollView: Tabs.ScrollView,
  FlatList: Tabs.FlatList,
  SectionList: Tabs.SectionList,
  Lazy: Tabs.Lazy,
  FlashList: Tabs.FlashList,
  MasonryFlashList: Tabs.MasonryFlashList
}

export { CustomTabs }
export * from 'react-native-collapsible-tab-view' // Export all props and types
