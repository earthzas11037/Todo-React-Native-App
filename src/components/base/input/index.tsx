import { inputTv } from '@/components/base/input/input.style'
import { Text } from '@/components/base/text'
import colors from '@/config/theme/colors'
import * as React from 'react'
import type { Control, FieldValues, Path, RegisterOptions } from 'react-hook-form'
import { useController } from 'react-hook-form'
import type { NativeSyntheticEvent, TextInputFocusEventData, TextInputProps } from 'react-native'
import { I18nManager, TextInput as NTextInput, StyleSheet, View } from 'react-native'
import { tv, VariantProps } from 'tailwind-variants'

type TextInputVariants = VariantProps<typeof inputTv>

export interface NInputProps extends TextInputVariants, Omit<TextInputProps, 'disabled'> {
  label?: string
  disabled?: boolean
  error?: boolean
  helperText?: string
  className?: string
  labelClassName?: string
  inputContainerClassName?: string
  inputClassName?: string
  helperTextClassName?: string
  startAdornment?: React.ReactNode
  endAdornment?: React.ReactNode
}

type TRule<T extends FieldValues> =
  | Omit<RegisterOptions<T>, 'disabled' | 'valueAsNumber' | 'valueAsDate' | 'setValueAs'>
  | undefined

export type RuleType<T extends FieldValues> = { [name in keyof T]: TRule<T> }
export type InputControllerType<T extends FieldValues> = {
  name: Path<T>
  control: Control<T>
  rules?: RuleType<T>
}

interface ControlledInputProps<T extends FieldValues> extends NInputProps, InputControllerType<T> {}

export const Input = React.forwardRef<NTextInput, NInputProps>((props, ref) => {
  const {
    label,
    color = 'primary',
    error,
    helperText,
    testID,
    placeholder,
    className,
    labelClassName,
    inputContainerClassName,
    inputClassName,
    helperTextClassName,
    disabled,
    startAdornment,
    endAdornment,
    onBlur: externalOnBlur,
    onFocus: externalOnFocus,
    ...inputProps
  } = props

  const id = React.useId()
  const [isFocussed, setIsFocussed] = React.useState(false)

  const handleBlur = React.useCallback(
    (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
      setIsFocussed(false)
      if (externalOnBlur) {
        externalOnBlur(e)
      }
    },
    [externalOnBlur]
  )

  const handleFocus = React.useCallback(
    (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
      setIsFocussed(true)
      if (externalOnFocus) {
        externalOnFocus(e)
      }
    },
    [externalOnFocus]
  )

  const styles = React.useMemo(
    () =>
      inputTv({
        error: Boolean(error),
        focused: isFocussed,
        disabled: Boolean(disabled),
        color
      }),
    [error, color, isFocussed, disabled]
  )

  return (
    <View className={styles.container({ className })}>
      {label && (
        <Text
          nativeID={`${id}-label`}
          accessibilityLabelledBy={id}
          testID={testID ? `${testID}-label` : undefined}
          className={styles.label({ className: labelClassName })}
        >
          {label}
        </Text>
      )}
      <View className={styles.inputContainer({ className: inputContainerClassName })}>
        {startAdornment && <View className='mr-2'>{startAdornment}</View>}
        <NTextInput
          accessibilityLabelledBy={`${id}-label`}
          accessibilityLabel={label}
          nativeID={id}
          testID={testID}
          ref={ref}
          placeholderTextColor={colors.text.disabled}
          placeholder={isFocussed ? undefined : placeholder}
          className={styles.input({ className: inputClassName })}
          onBlur={handleBlur}
          onFocus={handleFocus}
          selectionColor={colors.text.primary}
          autoComplete='off'
          autoCorrect={false}
          importantForAutofill='noExcludeDescendants'
          editable={!disabled}
          {...inputProps}
          style={StyleSheet.flatten([
            { writingDirection: I18nManager.isRTL ? 'rtl' : 'ltr' },
            { textAlign: I18nManager.isRTL ? 'right' : 'left' },
            inputProps.style
          ])}
        />
        {endAdornment && <View className='ml-2 justify-center'>{endAdornment}</View>}
      </View>
      {helperText && (
        <Text
          testID={testID ? `${testID}-helperText` : undefined}
          className={styles.helperText({ className: helperTextClassName })}
        >
          {helperText}
        </Text>
      )}
    </View>
  )
})

// only used with react-hook-form
export function ControlledInput<T extends FieldValues>(props: ControlledInputProps<T>) {
  const { name, control, rules, ...inputProps } = props

  const { field, fieldState } = useController({ control, name, rules })
  return (
    <Input
      ref={field.ref}
      autoCapitalize='none'
      onChangeText={field.onChange}
      value={(field.value as string) || ''}
      {...inputProps}
      error={!!fieldState.error}
      helperText={fieldState.error?.message}
    />
  )
}
