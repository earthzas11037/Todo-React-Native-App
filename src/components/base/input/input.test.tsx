import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react-native'
import { Text, TextInput } from 'react-native'
import { Input } from '@/components/base/input'

// Mock Adornment ง่ายๆ
const MockAdornment = ({ text }: { text: string }) => <Text>{text}</Text>

describe('Input (Presentational Component)', () => {
  // 1. การแสดงผลพื้นฐาน
  it('should render label, placeholder, and helper text correctly', () => {
    render(<Input label='Username' placeholder='Enter your username' helperText='This is a hint' />)
    expect(screen.getByText('Username')).toBeOnTheScreen()
    expect(screen.getByPlaceholderText('Enter your username')).toBeOnTheScreen()
    expect(screen.getByText('This is a hint')).toBeOnTheScreen()
  })

  // 2. การโต้ตอบของผู้ใช้
  it('should call onChangeText when user types', () => {
    const mockOnChangeText = jest.fn()
    render(<Input label='Test' onChangeText={mockOnChangeText} />)

    const input = screen.getByLabelText('Test') // ค้นหา Input จาก Label ที่เชื่อมกัน
    fireEvent.changeText(input, 'hello world')

    expect(mockOnChangeText).toHaveBeenCalledWith('hello world')
  })

  it('should call onFocus and onBlur callbacks', () => {
    const mockOnFocus = jest.fn()
    const mockOnBlur = jest.fn()
    render(<Input label='Focus Test' onFocus={mockOnFocus} onBlur={mockOnBlur} />)

    const input = screen.getByLabelText('Focus Test')

    fireEvent(input, 'focus')
    expect(mockOnFocus).toHaveBeenCalledTimes(1)

    fireEvent(input, 'blur')
    expect(mockOnBlur).toHaveBeenCalledTimes(1)
  })

  // 3. ทดสอบ Props ที่เป็นสถานะ
  it('should be non-editable when disabled', () => {
    render(<Input label='Disabled Test' disabled />)
    const input = screen.getByLabelText('Disabled Test')
    expect(input).toHaveProp('editable', false)
  })

  it('should display helper text when error prop is true', () => {
    render(<Input label='Error Test' error helperText='Invalid input' />)
    const helperText = screen.getByText('Invalid input')
    expect(helperText).toBeOnTheScreen()
    // คุณสามารถเช็ค style เพิ่มเติมได้ถ้าต้องการ แต่ snapshot จะครอบคลุมส่วนนี้
  })

  // 4. ทดสอบ Adornments
  it('should render start and end adornments', () => {
    render(
      <Input
        label='Adornment Test'
        startAdornment={<MockAdornment text='START' />}
        endAdornment={<MockAdornment text='END' />}
      />
    )
    expect(screen.getByText('START')).toBeOnTheScreen()
    expect(screen.getByText('END')).toBeOnTheScreen()
  })

  // 5. ทดสอบ Ref Forwarding
  it('should forward ref to the underlying TextInput', () => {
    const ref = React.createRef<TextInput>()
    render(<Input label='Ref Test' ref={ref} />)
    expect(ref.current).toBeTruthy()
  })

  // 6. Snapshot Testing สำหรับ Styles และ Variants
  describe('Snapshot for Variants', () => {
    it('renders correctly with error state', () => {
      const { toJSON } = render(<Input label='Error' error helperText='Error!' />)
      expect(toJSON()).toMatchSnapshot()
    })

    it('renders correctly when focused', () => {
      const { toJSON } = render(<Input label='Focused' />)
      // จำลองการ focus เพื่อให้ snapshot จับ state ที่ถูกต้อง
      fireEvent(screen.getByLabelText('Focused'), 'focus')
      expect(toJSON()).toMatchSnapshot()
    })

    it('renders correctly when disabled', () => {
      const { toJSON } = render(<Input label='Disabled' disabled />)
      expect(toJSON()).toMatchSnapshot()
    })
  })
})
