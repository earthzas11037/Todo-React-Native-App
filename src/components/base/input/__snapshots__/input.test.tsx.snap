// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input (Presentational Component) Snapshot for Variants renders correctly when disabled 1`] = `
<View>
  <Text
    accessibilityLabelledBy=":r9:"
    className="font-notoSansThai mt-[1px] font-normal mb-1 text-base text-text-disabled"
    nativeID=":r9:-label"
    style={
      {
        "writingDirection": "ltr",
      }
    }
  >
    Disabled
  </Text>
  <View
    className="rounded-lg px-4 py-2 min-h-[50px] flex flex-row shadow-none border bg-action-hover text-text-disabled border-outline-border"
  >
    <TextInput
      accessibilityLabel="Disabled"
      accessibilityLabelledBy=":r9:-label"
      autoComplete="off"
      autoCorrect={false}
      className="flex-1 font-notoSansThai text-[14px] font-normal leading-normal text-text-disabled"
      editable={false}
      importantForAutofill="noExcludeDescendants"
      nativeID=":r9:"
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#A0A6AD"
      selectionColor="#33383F"
      style={
        {
          "textAlign": "left",
          "writingDirection": "ltr",
        }
      }
    />
  </View>
</View>
`;

exports[`Input (Presentational Component) Snapshot for Variants renders correctly when focused 1`] = `
<View>
  <Text
    accessibilityLabelledBy=":r8:"
    className="font-notoSansThai mt-[1px] font-normal text-text-primary mb-1 text-base"
    nativeID=":r8:-label"
    style={
      {
        "writingDirection": "ltr",
      }
    }
  >
    Focused
  </Text>
  <View
    className="rounded-lg bg-white px-4 py-2 min-h-[50px] flex flex-row border-primary-500 border"
  >
    <TextInput
      accessibilityLabel="Focused"
      accessibilityLabelledBy=":r8:-label"
      autoComplete="off"
      autoCorrect={false}
      className="flex-1 font-notoSansThai text-[14px] font-normal leading-normal"
      editable={true}
      importantForAutofill="noExcludeDescendants"
      nativeID=":r8:"
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#A0A6AD"
      selectionColor="#33383F"
      style={
        {
          "textAlign": "left",
          "writingDirection": "ltr",
        }
      }
    />
  </View>
</View>
`;

exports[`Input (Presentational Component) Snapshot for Variants renders correctly with error state 1`] = `
<View>
  <Text
    accessibilityLabelledBy=":r7:"
    className="font-notoSansThai mt-[1px] font-normal mb-1 text-base text-error-500"
    nativeID=":r7:-label"
    style={
      {
        "writingDirection": "ltr",
      }
    }
  >
    Error
  </Text>
  <View
    className="rounded-lg bg-white px-4 py-2 min-h-[50px] flex flex-row shadow-none border border-error-500"
  >
    <TextInput
      accessibilityLabel="Error"
      accessibilityLabelledBy=":r7:-label"
      autoComplete="off"
      autoCorrect={false}
      className="flex-1 font-notoSansThai text-[14px] font-normal leading-normal"
      editable={true}
      importantForAutofill="noExcludeDescendants"
      nativeID=":r7:"
      onBlur={[Function]}
      onFocus={[Function]}
      placeholderTextColor="#A0A6AD"
      selectionColor="#33383F"
      style={
        {
          "textAlign": "left",
          "writingDirection": "ltr",
        }
      }
    />
  </View>
  <Text
    className="font-notoSansThai font-normal text-base text-error-500 mt-1"
    style={
      {
        "writingDirection": "ltr",
      }
    }
  >
    Error!
  </Text>
</View>
`;
